### DISCUSSION TITLE
Run Neovide on remote SSH system

### DISCUSSION THREAD

**Discussion by CGamesPlay** [neovide/neovide#2853] [Show and tell]
You can configure Neovide to connect to a headless nvim running on a remote server. Here are some methods to do it.

## NVRH

You can use the [nvrh](https://github.com/mikew/nvrh) (Neovim Remote Helper) tool to do this. Nvrh needs to be installed on the same machine as Neovide but only SSH and nvim need to be installed on the remote machine.

```bash
nvrh client open \
    --local-editor /Applications/Neovide.app/Contents/MacOS/neovide \
    --local-editor --server \
    --local-editor {{SOCKET_PATH}} \
    --local-editor --no-fork \
    my-host \
    remote/path/to/project
```

NVRH also adds some nvim commands to add port forwards to your SSH connection while it's running and sets up the URL handler in nvim so that links open in your local browser.

NVRH (v0.1.20) doesn't currently allow setting a custom command, so you cannot use this method to access a remote devcontainer.

## Manual / customizable version, Neovide 0.15.0+

You can do this manually using SSH directly. The trick is to set Neovide's `--neovim-bin` argument to launch a remote SSH editing session. You probably want to use a wrapper script to do it:

```bash
#!/bin/bash
# Save as connect-to-myhost.sh
# You can add extra SSH arguments, like port forwards
exec ssh myhost -- cd remote/directory ";" nvim "$@"
```

Then run Neovide: `neovide --wsl --neovim-bin=/path/to/connect-to-myhost.sh` If you are on macOS, you can use `open -na Neovide --args --wsl --neovim-bin=/path/to/connect-to-myhost.sh`.

The `--wsl` flag causes Neovide to set a custom clipboard handler, which is required for WSL (Windows Subsystem for Linux), and also for when nvim is running on a remote server. Thanks @TwIStOy for discovering this.

### Making a generic wrapper

Neovide will pass all arguments after its own on to the neovim-bin script, so your wrapper script can take its own arguments as well. 

```
#!/bin/bash

args=()
for arg in "$@"; do
    case "$arg" in
        --prefix=*)
            prefix=${arg#--prefix=}
            ;;
        *)
            args+=("$arg")
            ;;
    esac
done
exec sh -c "$prefix $(printf '%q ' "${args[@]}")"
```

This one now takes a shell script to run. Use it like this: `neovide --wsl --neovim-bin=/path/to-wrapper.sh -- --prefix="ssh myhost -- cd remote/directory; nvim"`

## Old, pre-0.15.0 version
<details><summary>Details</summary>
<p>

Just found out that you can use `--neovim-bin` to be *any command*, so you can easily launch a remote SSH editing session:

```
neovide --neovim-bin="ssh myhost -- cd remote/directory; nvim"
```

You can take this a step further and use a remote devcontainer with docker:

```
neovide --neovim-bin="ssh myhost -- docker exec -i mycontainer nvim"
```

</p>
</details>

**Comment by masoomshaikh** [comment]
Nifty, found while looking for same, unfortunately doesn't work on Windows!

I have to fallback to using below setup
On Remote
```
nvim --headless --listen 10.10.10.10:1234
```
On Local
```
neovide --server 10.10.10.10:1234
```

change your IP address and port accordingly and adjust your firewalls!

**Reply by CGamesPlay** [reply]
Definitely don't adjust your firewalls! The connection is unsecured and unauthenticated and would give anyone access to run arbitrary commands as nvim's user. You can use SSH with the server mode as well, I just found it less convenient:

```
ssh myhost -L 1234:localhost:1234 -- nvim --headless --listen localhost:1234
# In another terminal
neovide --server localhost:1234
```

**Reply by Strackeror** [reply]
You can actually make it work on windows, you just have to collapse the argument forwarding logic to a single executable. Here's a little rust program as an example: 
```rust
use std::env::args;

fn main() {
    let basic_args = ["-t", "vm", "--", "nvim"];
    std::process::Command::new("ssh")
        .args(basic_args)
        .args(args().skip(1))
        .status()
        .unwrap();
}
```

It's a lot more comfortable to use that way, since on the windows side I can just have a shortcut to `neovide.exe --neovim-bin=ssh_launcher.exe` and I can open as many windows as needed without having to worry about port collisions or extra command windows.

**Comment by Trantorian1** [comment]
Following up on this thread, here is the command I am using to set the directory and config location on the remote machine:

```bash
ssh myhost -L 1234:localhost:1234 -- "cd path/to/project; path/to/nvim -u ~/path/to/init.lua --headless --listen localhost:1234"
# In another terminal
neovide --server localhost:1234
```

I'm sure there's a nicer way to automate connecting to a remote session but this works fine atm. Strangely enough it does not seem to be possible to set neovide options once in the remote session.

**Reply by sawyerzheng** [reply]
Single command to start remote ssh nvim and neovide frontend:

windows bat file start-neovide.bat(tested):

``` bat
for /f %%i in ('python -c "import random;print(random.randrange(30000,64000))"') do set random_number=%%i

echo %random_number%
start /b ssh %1 -v -L %random_number%:localhost:%random_number% -- nvim --headless --listen localhost:%random_number%

:: sleep 1s to wait nvim headless server starting
timeout /t 1 /nobreak >nul 


neovide.exe --server=localhost:%random_number%
```

linux startneovide.sh

``` bash
#!/bin/bash

# Generate random port number between 30000 and 64000
random_number=$(python -c "import random; print(random.randrange(30000, 64000))")
echo $random_number

# Start SSH with port forwarding and nvim in background
ssh "$1" -v -L "$random_number:localhost:$random_number" -- nvim --headless --listen "localhost:$random_number" &

# sleep to wait nvim headless server starting
sleep 1

# Start neovide connecting to the remote nvim instance
neovide --server="localhost:$random_number"
```

Usage `start-neovide.cmd your-ssh-server` or `startneovide.sh your-ssh-server`.

**Reply by CGamesPlay** [reply]
On multi-user systems this is still unsafe, as any user on the system can effectively log in as your user without authentication using that nvim instance. However, since these are listening on localhost, it's not a network-level vulnerability.

I haven't gotten any feedback on why my script in the OP is less convenient, but it's secure since it doesn't use nvim's server mode. I guess people just don't know / don't care about the security aspect, but I'm happy to address any usability concerns.

**Reply by DeadlySquad13** [reply]
Had issues connecting via nvrh (Windows -> MacOs): the connection would be closed before neovide has opened while cli nvim worked just fine. But the bat file works, thanks! I guess it's the trick with a sleep.

UPD: Updated version of a neovide from 0.14.0 to the 0.15.2 and now everything
works with nvrh… Maybe `--no-fork` was missing in my version… Thanks to all participants in this thread! The whole story with remote connection is a bit more convoluted than I might have thought.

**Comment by CGamesPlay** [comment]
I've been using this daily since I originally posted. The biggest drawback is the lack of clipboard support. This could be resolved by Neovide providing the clipboard handler instead of relying on nvim's builtin one.

**Reply by SuperBo** [reply]
thank you, I also use your solution heavily recently. I found this approach more convenient than nvim server and client approach.

**Reply by colawithsauce** [reply]
for clipboard support, you can have a try with my solution... https://github.com/colawithsauce/clipboard_server

**Reply by TwIStOy** [reply]
About the clipboard, you can try to add `--wsl` to your cmdline. Although you are not connecting a neovim instance in wsl. But it will make neovide to use your local clipboard. It's a hack. But it works for me.

**Reply by CGamesPlay** [reply]
Wow, neat trick. Doesn't seem to work for me on macOS, but I assume you are using Windows. Still, that same code path might be the way to properly add support for this to Neovide.

**Reply by TwIStOy** [reply]
emmm. It works for me on macOS.
https://github.com/neovide/neovide/blob/main/src/bridge/mod.rs#L102
They only check if wsl is enalbed then handle clipbaord.

**Reply by CGamesPlay** [reply]
I just got a chance to try this again and you're totally right. I wasn't properly setting the flag in my activation script. I've updated my initial directions to include this flag.

Great find!

**Comment by CGamesPlay** [comment]
With Neovide 0.15.0, the way that you use this technique has changed, but it still works. I updated my top-level post with new instructions.

**Comment by mikew** [comment]
I think [nvrh](https://github.com/mikew/nvrh) is basically this? You can do

```bash
export NVRH_CLIENT_LOCAL_EDITOR=/Applications/Neovide.app/Contents/MacOS/neovide,--server,{{SOCKET_PATH}},--no-fork

nvrh client open some.ip.address.yeah path/to/project
```

**Reply by CGamesPlay** [reply]
Nice project! I've added it to my post. Feature request: allow customizing the remote command (and maybe SSH options). As it stands, you can't use this with a devcontainer. To use a devcontainer over SSH, you'd use this kind of command:

```bash
ssh myhost -- npx devcontainers/cli exec --workspace-folder /path/to/devcontainer -- nvim "$@"
```

**Reply by mikew** [reply]
Neat, was wondering how dev containers would work, customizing the nvim command is pretty easy and I surprised myself by missing it.

what ssh options are you thinking? Could they not be set in `.ssh/config` (genuinely asking)?

Also, there's [remote-nvim](https://github.com/amitds1997/remote-nvim.nvim) which seems pretty impressive.

**Reply by CGamesPlay** [reply]
I use my own [qtm](https://gitlab.com/CGamesPlay/qtm) project to implement port forwards in devcontainers, and it also allows me to set up a "durable SSH session" like Eternal Terminal. To get that behavior, I use `ssh myhost -o Hostname=localhost -p 2222`, and I *could* put this in `~/.ssh/config`, but I actually want the option to use normal SSH.

[append]
I've thought about this and realized that the port forwards you are using are incompatible with devcontainers anyways (because devcontainers need to expose their ports to be accessible by the SSH host and this cannot be done dynamically), so maybe this feature request doesn't make sense.

**Comment by michaelfortunato** [comment]
So I guess if we connect to a remote with ssh and provide the -X flag it should be possible to render plots in a gui window ?

**Reply by CGamesPlay** [reply]
If that normally works over ssh without Neovide involved, then it should work with Neovide being the one to run ssh.

**Comment by ollien** [comment]
Since version 0.16, you need to create a wrapper script for this. Here's an example of the script I use

```sh
#!/usr/bin/env bash
exec ssh devbox -- "/opt/nvim/bin/nvim $@"
```

And then in the neovide config

```toml
neovim-bin="/path/to/bin/remote-nvim"
```
